alephium/contracts/token_bridge/token_bridge_for_chain.ral

```diff
@@ -0,0 +1,69 @@
+    mut next: U256,
+    mut next1: U256,
+    mut next2: U256,
+    undoneSequenceTemplateId: ByteVec, // used for create contract by copyCreateContract
+    refundAddress: Address,
+    mut sendSequence: U256
+) extends Sequence(next, next1, next2, undoneSequenceTemplateId, refundAddress), Constants(), TokenBridgeConstants() {
+    pub fn nextSendSequence() -> U256 {
+        checkPermission!(callerContractId!() == localTokenBridgeId, ErrorCodes.InvalidCaller)
+        let msgSequence = sendSequence
+        sendSequence = sendSequence + 1
+        return msgSequence
+    }
+
+    fn checkTransfer(vaa: ByteVec) -> (Bool, ByteVec) {
+        let governance = Governance(governanceContractId)
+        let (emitterChainId, targetChainId, emitterAddress, msgSequence, payload) = governance.parseAndVerifyVAA(vaa)
+        assert!(emitterChainId == remoteChainId, ErrorCodes.InvalidEmitChainId)
+        assert!(emitterAddress == remoteTokenBridgeId, ErrorCodes.InvalidEmitAddress)
+        assert!(targetChainId == localChainId, ErrorCodes.InvalidTargetChainId)
+        return checkSequence(msgSequence), payload
+    }
+
+    @using(readonly = true)
+    fn parseCompleteTransfer(payload: ByteVec) -> (U256, ByteVec, U256, Address, U256) {
+        assert!(byteVecSlice!(payload, 0, 1) == PayloadId.Transfer, ErrorCodes.InvalidPayloadId)
+        let amount = u256From32Byte!(byteVecSlice!(payload, 1, 33))
+        let vaaTokenId = byteVecSlice!(payload, 33, 65)
+        let vaaTokenChainId = u256From2Byte!(byteVecSlice!(payload, 65, 67))
+        let toAddress = byteVecSlice!(payload, 67, 99)
+        let normalizedArbiterFee = u256From32Byte!(byteVecSlice!(payload, 99, 131))
+        assert!(size!(payload) == 131, ErrorCodes.InvalidMessageSize)
+        let recipient = byteVecToAddress!(#00 ++ toAddress)
+        return amount, vaaTokenId, vaaTokenChainId, recipient, normalizedArbiterFee
+    }
+
+    @using(preapprovedAssets = true)
+    pub fn completeTransfer(vaa: ByteVec, caller: Address) -> () {
+        let (sequenceReady, payload) = checkTransfer(vaa)
+        if (!sequenceReady) { // too many undone sequences in between, compact those sequences first
+            return
+        }
+
+        let (amount, vaaTokenId, vaaTokenChainId, recipient, normalizedArbiterFee) = parseCompleteTransfer(payload)
+        let path = Path.TokenPool ++ u256To2Byte!(vaaTokenChainId) ++ vaaTokenId
+        let tokenPoolId = subContractIdOf!(TokenBridge(localTokenBridgeId), path)
+        let tokenPool = TokenPool(tokenPoolId)
+        tokenPool.completeTransfer{caller -> dustAmount!()}(
+            remoteChainId, amount, vaaTokenId, vaaTokenChainId, recipient, normalizedArbiterFee, caller
+        )
+    }
+
+    pub fn destroyUndoneSequenceContracts(paths: ByteVec) -> () {
+        checkPermission!(callerContractId!() == localTokenBridgeId, ErrorCodes.InvalidCaller)
+        let length = size!(paths)
+        let mut index = 0
+        while (index < length) {
```

Comment:
Question: what is the use case for this command?

If this is executed successfully, what is the consequence for the `Sequence` contract? Would the `checkSequence` method still work?

Reply:
`UndoneSequence` contract should not or rarely be created in the normal case, but if we are attacked, it may create lots of `UndoneSequence` contracts, so we add a governance action to delete these contracts.

Reply:
Yeah, got it. I was thinking about the legitimate ones, like in this comment: https://github.com/alephium/wormhole-fork/pull/5#discussion_r958259375

So basically all the "good" ones during the attack will be invalid, which could be an attack if the attacker knows that there is a huge tranfer (10M dollars)  and just flood the system with very small transfers to sabotage that.

Reply:
In this case, we have to wait for the normal user to execute withdraw before destroy the undone sequence contracts.